### PR TITLE
Add Scheduler wrapper and proto for scheduled rules

### LINKED ISSUE #86: Create service class for Cloud Scheduler

### PR DESCRIPTION
**Pull request by thu5**
This PR adds a wrapper class for Cloud Scheduler. It also creates a proto to represent scheduled rules that will be used by Scheduler. Closes #86.

### CODE REVIEW

File: communication.proto [hunk 1]
```diff
@@ -0,0 +1,14 @@
+syntax = "proto2";
```
Comment (gg-qinp) [inline communication.proto]: the proto file name can be more specific.

File: server/lib/service/scheduler.dart [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  ///
+  /// Throws an [ApiRequestError] if Scheduler API returns an error.
+  Future<void> _createJob(
+      String name,
+      String service,
+      String path,
+      String cronExpression,
+      String method,
+      Map<String, String> headers,
+      List<int> body,
```
Comment (gg-qinp) [inline server/lib/service/scheduler.dart:73]: is this the http post request body which is sent to DV360 Public API server? it only contains userId and ruleId, wondering is it enough

File: communication.proto (hunk 1 above)
Comment (thu5) [reply communication.proto]: Ok 🙂

File: server/lib/service/scheduler.dart (hunk 2 above)
Comment (thu5) [reply server/lib/service/scheduler.dart:73]: No, this body is sent back to the App Engine server. The App Engine server will then look up the rule information in Firestore and then use the DV360 Public API to make changes.

File: scheduled_rule.proto [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+package dv360_excel_plugin;
+
+// A message that represents a scheduled rule to be sent to Cloud Scheduler.
+message ScheduledRule {
+  // required
+  // The user identifier.
+  optional string user_id = 1;
+
+  // required
+  // The rule identifier.
```
Comment (cathyang-work) [inline scheduled_rule.proto]: Maybe mention generated by Firestore?

File: server/lib/service/scheduler.dart [hunk 4]
```diff
@@ -0,0 +1,87 @@
+import 'package:googleapis/cloudscheduler/v1.dart';
+import 'package:http/http.dart';
+
+import '../proto/scheduled_rule.pb.dart';
+import '../proto/rule.pb.dart';
+
+/// A class that wraps around Google Cloud Scheduler.
+class SchedulerClient {
+  /// The target HTTP endpoint for each job.
```
Comment (cathyang-work) [inline server/lib/service/scheduler.dart]: targeted?

File: server/lib/service/scheduler.dart [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    final ruleDetails = ScheduledRule()
+      ..ruleId = rule.id
+      ..userId = userId;
+
+    // Depending on the schedule type, creates the Scheduler job.
+    if (rule.schedule.type == Schedule_Type.REPEATING) {
+      final schedule = rule.schedule.repeatingParams.cronExpression;
+      await _createJob(jobName, _appEngineService, path, schedule, method,
+          headers, ruleDetails.writeToBuffer(), timezone);
+    }
```
Comment (cathyang-work) [inline server/lib/service/scheduler.dart:56]: Suggest to throw UnimplementedError if it is other Schedule_Type

```
throw UnimplementedError('...');
```
https://api.dart.dev/stable/2.9.1/dart-core/UnimplementedError-class.html

File: scheduled_rule.proto [hunk 6]
```diff
@@ -0,0 +1,14 @@
+syntax = "proto2";
+package dv360_excel_plugin;
+
+// A message that represents a scheduled rule to be sent to Cloud Scheduler.
+message ScheduledRule {
+  // required
+  // The user identifier.
```
Comment (cathyang-work) [inline scheduled_rule.proto]: Maybe mention it is the obfuscated gaia id.

File: server/lib/service/scheduler.dart (hunk 5 above)
Comment (thu5) [reply server/lib/service/scheduler.dart:56]: Okay, that makes sense

File: server/lib/service/scheduler.dart [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  final String _appEngineService;
+
+  /// Creates an instance of [SchedulerClient].
+  SchedulerClient(Client client, this._projectId, this._locationId,
+      this._appEngineService, String baseUrl)
+      : _api = CloudschedulerApi(client, rootUrl: baseUrl);
+
+  /// Schedules a rule using Cloud Scheduler.
+  ///
+  /// Throws an [ApiRequestError] if Scheduler API returns an error.
```
Comment (cathyang-work) [inline server/lib/service/scheduler.dart]: Update comment as well.

**Review by cathyang-work — APPROVED** [review]

File: server/lib/service/scheduler.dart (hunk 5 above)
Comment (cathyang-work) [reply server/lib/service/scheduler.dart:56]: Can you also add a unit test for it?

**Review by gg-qinp — APPROVED** [review]

**Pull request merged**